policybot/dashboard/oauth.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dashboard
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+)
+
+type loginHandler struct {
+	clientID    string
+	secretState string
+}
+
+type callbackHandler struct {
+	clientID     string
+	clientSecret string
+	secretState  string
+	rc           RenderContext
+}
+
+// Support for GitHub oauth flow
+func newOAuthHandlers(clientID string, clientSecret string, rc RenderContext) (login http.Handler, callback http.Handler) {
+	// secret state for OAuth exchanges
+	secretState := make([]byte, 32)
+	_, _ = rand.Read(secretState)
+
+	ss := base64.StdEncoding.EncodeToString(secretState)
+
+	return &loginHandler{
+			clientID:    clientID,
+			secretState: ss,
+		}, &callbackHandler{
+			clientID:     clientID,
+			secretState:  ss,
+			clientSecret: clientSecret,
+			rc:           rc,
+		}
+}
+
+func (h *loginHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	url := "https://github.com/login/oauth/authorize?client_id=" + h.clientID + "&scope=user,repo&state=" + h.secretState
+	http.Redirect(w, r, url, http.StatusTemporaryRedirect)
+}
+
+func (h *callbackHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	httpClient := http.Client{}
+
+	if err := r.ParseForm(); err != nil {
+		h.rc.RenderHTMLError(w, fmt.Errorf("unable to parse query: %v", err))
+		return
+	}
+
+	if r.FormValue("state") != h.secretState {
+		h.rc.RenderHTMLError(w, fmt.Errorf("unable to verify request state"))
+		return
+	}
+
+	url := fmt.Sprintf("https://github.com/login/oauth/access_token?client_id=%s&client_secret=%s&code=%s", h.clientID, h.clientSecret, r.FormValue("code"))
```

Comment:
Do these need to be escaped?

Reply:
I think those strings are inherently safe for URL encoding, but it's wise to do it anyway for the sake of being more robust...